workspace_file_harvester/app.py

```
@@ -0,0 +1,169 @@
+import asyncio
+import datetime
+import hashlib
+import json
+import logging
+import os
+import traceback
+import uuid
+
+import boto3
+from botocore.exceptions import ClientError
+from eodhp_utils.aws.s3 import upload_file_s3
+from eodhp_utils.runner import get_boto3_session, get_pulsar_client, setup_logging
+from fastapi import FastAPI
+from messager import FileHarvesterMessager
+from starlette.responses import JSONResponse
+
+root_path = os.environ.get("ROOT_PATH", "/")
+logging.error("Starting FastAPI")
+app = FastAPI(root_path=root_path)
+
+source_s3_bucket = os.environ.get("SOURCE_S3_BUCKET")
+target_s3_bucket = os.environ.get("TARGET_S3_BUCKET")
+
+setup_logging(verbosity=2)
+
+minimum_message_entries = int(os.environ.get("MINIMUM_MESSAGE_ENTRIES", 100))
+
+
+def get_file_hash(data: str) -> str:
+    """Returns hash of data available"""
+
+    def _md5_hash(byte_str: bytes) -> str:
+        """Calculates an md5 hash for given bytestring"""
+        md5 = hashlib.md5()
+        md5.update(byte_str)
+        return md5.hexdigest()
+
+    return _md5_hash(data.encode("utf-8"))
+
+
+def get_file_s3(bucket: str, key: str, s3_client: boto3.client) -> tuple:
```

The interface this function provides is roughly:

> Read a UTF8 text file from an S3 bucket and return a string and its last-modified date. If the file doesn't exist, is not valid UTF8, or any kind of error occurs, including rate limits and network errors, then return an empty dictionary instead.

This presents a couple of small problems:

* The type of its return value is either `Tuple[str, datetime]` or `Tuple[dict, datetime]`, which is difficult to deal with. Indeed, its only caller treats it as [str, datetime] and then catches the TypeError which is not just a bit more complex than necessary but risks masking a genuine TypeError.
* The error handling behaviour isn't very useful for its caller. It might as well let the exception propagate and have the caller catch (replacing the `except TypeError`).
* It's clearly designed and only really usable for a specific purpose - getting the metadata file - but its name and docstring suggest its more widely applicable.

It'd be simpler and clearer to combine this into `get_metadata`

---

Merged functions